[21363] Migration guide (#5102)
eProsima/Fast-DDS: UPGRADING.md

* CMake Windows file names have been changed:

    * fastdds.manifest.in
    * fastdds-config.cmake

* The fastrtps.rc file has been renamed as fastdds.rc.
* Deprecated APIs marked with `FASTDDS_DEPRECATED` and `FASTDDS_TODO_BEFORE` macros have been removed.

## Compatibility with Fast CDR

Fast DDS v3.x is only compatible with Fast CDR v2.x.

## Namespace migrations and changes

The following list contains the namespace changes and migrations:

* All `eprosima::fastrtps::` namespace references were moved to `eprosima::fastdds::`.
* `SubscriptionBuiltinTopicData`, `PublicationBuiltinTopicData` and `ParticipantBuiltinTopicData` were moved from `fastdds::dds::builtin::` to `fastdds::dds::`.
* `EventKindBits::` references changed to`EventKind::`.
* `EventKindEntityId::` references changed to`EntityId::`.
* `StatisticsEventKind::` references changed to `statistics::EventKind::`.
* `Duration_t` and `c_TimeInfinite` reference were moved to `dds::`.
* `ParticipantAttributes`, `PublisherAttributes`, `SubscriberAttributes`, `ReplierAttributes`, `RequesterAttributes` were moved to `xmlparser::`.
* `Time_hpp` references were divided into `dds::Time_t::` and `rtps::Time_t::`.

> 

Reviewer: ```suggestion
```

| fastrtps/qos/LivelinessChangedStatus.h | fastdds/dds/core/status/LivelinessChangedStatus.hpp |
| fastrtps/qos/QosPolicies.h | fastdds/dds/core/policy/QosPolicies.hpp |
| fastrtps/qos/ReaderQos.h | fastdds/dds/subscriber/qos/ReaderQos.hpp |
| fastrtps/qos/WriterQos.h | fastdds/dds/publisher/qos/WriterQos.hpp |
| fastrtps/qos/SampleRejectedStatus.hpp | fastdds/dds/core/status/SampleRejectedStatus.hpp |
| fastrtps/participant/Participant.h | fastdds/rtps/participant/RTPSParticipant.hpp |
| fastrtps/transport/TCPv4TransportDescriptor.h | fastdds/rtps/transport/TCPv4TransportDescriptor.hpp |
| fastrtps/transport/TCPv6TransportDescriptor.h| 	fastdds/rtps/transport/ TCPv6TransportDescriptor.hpp |
| fastrtps/transport/UDPv4TransportDescriptor.h | fastdds/rtps/transport/ UDPv4TransportDescriptor.hpp |
| fastrtps/transport/UDPv6TransportDescriptor.h | fastdds/rtps/transport/ UDPv6TransportDescriptor.hpp |
| fastrtps/transport/UDPTransportDescritpor.h | fastdds/rtps/transport/UDPTransportDescritpor.hpp |
| fastrtps/transport/TCPTransportDescritpor.h | fastdds/rtps/transport/TCPTransportDescritpor.hpp |

> 

Reviewer: ```suggestion
```

| TopicDataType::deleteData(void*) | TopicDataType::delete_data(void*) |
| TopicDataType::getKey(const void* const, `InstanceHand`*, bool) | TopicDataType::compute_key(const void* const, `InstanceHand`&, bool) |
| TopicDataType::setName(const char*) | TopicDataType::set_name(const string&) |
| char* TopicDataType::getName() | string& TopicDataType::get_name() |
| TypeSupport::calculate_serialized_size_provider(const void* const, `DataRepresentationId_t`) | TypeSupport::calculate_serialized_size(const void* const, `DataRepresentationId_t`) |
|   get_key(void, `InstanceHandle_t`*, bool) | compute_key(`SerializedPayload_t`&, `InstanceHandle_t`&, bool) |
| DynamicPubSubType::createData() | DynamicPubSubType::create_data() |
| DynamicPubSubType::deleteData(void*) | DynamicPubSubType::delete_data(void*) |
| DynamicPubSubType::getKey(const void* const, `InstanceHand`*, bool) | DynamicPubSubType::compute_key(const void* const, `InstanceHand`&, bool) |
| DynamicPubSubType::getSerializedSizeProvider(const void* const, `DataRepresentationId_t`) | DynamicPubSubType::calculate_serialized_size(const void* const, `DataRepresentationId_t`) |

> 
> 

Reviewer: ```suggestion
```

## Struct, Enum, Variable

> 

Reviewer: Add some link sentence or small paragraph here

* Extend `SubscriptionBuiltinTopicData` with additional fields to mimic those of `ReaderProxyData`.
* Extend `PublicationBuiltinTopicData` with additional fields to mimic those of `WriterProxyData`.
* Extend `ParticipantBuiltinTopicData` adding the the Product version and additional fields to mimic those of `ParticipantProxyData`.
* `DiscoveryProtocol_t` is `DiscoveryProtocol`.
* Extend `SendBuffersAllocationAttributes` with a new attribute defining the allocation configuration of the `NetworkBuffers`.
* `TypeConsistencyQos` was removed from DataReader, and `TypeConsistencyEnforcementQosPolicy` and `DataRepresentationQosPolicy` were included.
* `initialHeartbeatDelay` is `initial_heartbeat_delay`.
* `heartbeatPeriod` is `heartbeat_period`.
* `nackResponseDelay` is `nack_response_delay`.
* `nackSupressionDuration` is `nack_supression_duration`.
* `heartbeatResponseDelay` is `heartbeat_response_delay`.
* `initialAcknackDelay` is `initial_acknack_delay`.
* `expectsInlineQos` is `expects_inline_qos`.
* `m_typeSize` is `max_serialized_type_size`
* `m_isGetKeyDefined` is `is_compute_key_provided`
* `m_topicDataTypeName` is `topic_data_typename_`

> ## Examples
> 

Reviewer: ```suggestion
## Examples

All the examples have been refactored to follow the same structure:
* File names, guards, and classes follow new format.
* Detailed and well-formed README.md with example explanation.
* Example structured in applications, stopped by `SIGTERM` signal.
```

### Hello World
Refactor the HelloWorld example with the current new example format.
In this hello world example, the key changes are:
* Removed --env option and set that environment behavior as default.
* Added subscriber waitsets class and its usage.
* Provided 	XML profiles examples that target several scenarios (e.g.,SampleConfig_Controller, Events, Multimedia).

### X-Types Examples

> In this X-Types example, a type is defined at runtime on the publisher side using the Dynamic Types API, and the subscriber discovers the type, creates a reader for it, and prints the received data. This example is type compatible with the Hello World example, and a compatibility test has been added for this.

Reviewer: ```suggestion
In this X-Types example, a type is defined at runtime on the publisher side using the Dynamic Types API, and the subscriber discovers the type, creates a reader for it, and prints the received data. 
This example is type compatible with the Hello World example, and a compatibility test has been added for this.
```

### Configuration
In this configuration example, the key changes are:
* Included LargeData as an option (builtin transport argument).
* Included all previous QoS examples:
    * Deadline
    * Disable positive ACKs
    * Lifespan
    * Liveliness
    * Ownership (strength)

>     * Partitions

Reviewer: Try to write this as a parragraph, as the suggestion in helloworld example. This applies also to the remain examples

Author: Don't you think it's clearer as a bullet list?

### Content Filter
Refactor the ContentFilteresTopicExample example with the current new example format.
In this content filter example, the main changes are:
* Added	option to select filter type: Default, Custom, or None.
* Customizable lower-bound and upper-bound options of the filter as arguments.
    * For the Custom filter, they represent the maximum and minimum values of the message indexes that are filtered out through the filter.
    * For the Default filter, they represent the maximum and minimum value message indexes that are read.

### Custom Payload Pool
Refactor the CustomPayloadPoolExample example with the current new example format.

### Delivery Mechanisms
In this delivery mechanisms example, the key changes are:
* Loans and data-sharing compatible: bounded types, final extensibility.
* Loans mechanism for data management.
* Option to select all delivery mechanisms.

### Discovery Server
Refactor the DiscoveryServerExample example with the current new example format.

> This refactors the DiscoveryServerExample to the new example format. It also adds proper tests for the examples testing automation.

Reviewer: ```suggestion
This refactors the DiscoveryServerExample to the new example format. 
It also adds proper tests for the examples testing automation.
```

# Migration Guide from Fast DDS v2 to Fast DDS v3

This document aims to help during the migration process from eProsima *Fast DDS version* 2 to *Fast DDS version* 3.
For more information about all the updates, please refer to the [release notes](https://fast-dds.docs.eprosima.com/en/latest/notes/notes.html).

> It is always *required* to regenerate the types with the latest version of *eProsima Fast DDS Gen*, *Fast DDS Gen* v4.0.0.

Reviewer 2: ```suggestion
It is always *required* to regenerate the types with the appropriate version of *eProsima Fast DDS Gen* (see [products compatibility](https://fast-dds.docs.eprosima.com/en/latest/notes/versions.html#eprosima-products-compatibility))
```

The following sections describe the possible changes that your project may require to migrate to *Fast DDS v3.0.0*.

- [Library management](#library-management)
- [Compatibility with Fast CDR](#compatibility-with-fast-cdr)
- [Namespace migrations and changes](#namespace-migrations-and-changes)
- [Public headers migrated to *fastdds*](#public-headers-migrated-to-fastdds)
- [Public headers moved to private](#public-headers-moved-to-private)
- [API changes](#api-changes)
- [Struct, Enum, Variable](#struct-enum-variable)
- [Examples](#examples)

> 

Reviewer 2: ```suggestion
```

## Library management

> The below exposes exposes the changes related to the package name, environment variables and other library usages.

Reviewer 2: ```suggestion
The list below exposes the changes related to the package name, environment variables and other library usages.
```

* Deprecated APIs marked with `FASTDDS_DEPRECATED` and `FASTDDS_TODO_BEFORE` macros have been removed.

## Compatibility with Fast CDR

Fast DDS v3 is only compatible with Fast CDR v2.
If you are not using Fast CDR as [third-party](https://fast-dds.docs.eprosima.com/en/latest/installation/configuration/cmake_options.html#third-party-libraries-options), please ensure that your local dependencies are up-to-date.

## Namespace migrations and changes

The following list contains the namespace changes and migrations:

* All `eprosima::fastrtps::` namespace references were moved to `eprosima::fastdds::`.
* `SubscriptionBuiltinTopicData`, `PublicationBuiltinTopicData` and `ParticipantBuiltinTopicData` were moved from `fastdds::dds::builtin::` to `fastdds::dds::`.
* `EventKindBits::` references changed to`EventKind::`.
* `EventKindEntityId::` references changed to`EntityId::`.
* `StatisticsEventKind::` references changed to `statistics::EventKind::`.
* `Duration_t` and `c_TimeInfinite` references were moved to `dds::`.
* `Time_t.hpp` references were moved from `eprosima::fastdds::` to `eprosima::fastdds::dds`.

## Public headers migrated to *fastdds*

> All the extensions of the headers in /include (public headers) have been changed to `.hpp`.

Reviewer 2: ```suggestion
All the extensions of the headers under `include`, that is public headers that applications can include, have been changed to `.hpp`.
```

# Migration Guide from Fast DDS v2 to Fast DDS v3

This document aims to help during the migration process from eProsima *Fast DDS version* 2 to *Fast DDS version* 3.
For more information about all the updates, please refer to the [release notes](https://fast-dds.docs.eprosima.com/en/latest/notes/notes.html).

> As announced in previous minor releases, it is always *recommended* to regenerate the types with the latest *eProsima Fast DDS Gen* release.

Reviewer 2: It's not always the latest! It's the latest patch associated with the corresponding Fast DDS version

In this case, the type regeneration is required, using *Fast DDS Gen* v4.0.0.

The following sections describe the possible changes that your project may require to migrate to *Fast DDS v3.0.0*.

- [Library management](#library-management)
- [Compatibility with Fast CDR](#compatibility-with-fast-cdr)
- [Namespace migrations and changes](#namespace-migrations-and-changes)
- [Public headers migrated to *fastdds*](#public-headers-migrated-to-fastdds)
- [Public headers moved to private](#public-headers-moved-to-private)
- [API changes](#api-changes)
- [Struct, Enum, Variable](#struct-enum-variable)
- [Examples](#examples)

## Library management

> The below list expose the changes related to the package name, environment variables and other library usages.

Reviewer 2: ```suggestion
The list below exposes the changes related to the package name, environment variables and other library usages.
```

> * The package has been renamed from `fastrtps` to `fastdds`.

Reviewer 2: ```suggestion
* The CMake project has been renamed from `fastrtps` to `fastdds`.
```

* XML profiles loading environment variable has been renamed to: `FASTDDS_DEFAULT_PROFILES_FILE`.
* The configuration file that Fast DDS looks for to load the profiles has been renamed to `DEFAULT_FASTDDS_PROFILES.xml`.

> * XML Schema namespace in fastdds_profiles.xsd has been updated to http://www.eprosima.com.

Reviewer 2: ```suggestion
* XML Schema namespace in `fastdds_profiles.xsd` has been updated to http://www.eprosima.com.
```

* CMake Windows file names have been changed:

    * fastdds.manifest.in
    * fastdds-config.cmake

* The fastrtps.rc file has been renamed as fastdds.rc.
* Deprecated APIs marked with `FASTDDS_DEPRECATED` and `FASTDDS_TODO_BEFORE` macros have been removed.

## Compatibility with Fast CDR

Fast DDS v3 is only compatible with Fast CDR v2.

> If you are not using Fast CDR as [third-party](https://fast-dds.docs.eprosima.com/en/latest/installation/configuration/cmake_options.html#third-party-libraries-options), please ensure updating your local dependencies.

Reviewer 2: ```suggestion
If you are not using Fast CDR as [third-party](https://fast-dds.docs.eprosima.com/en/latest/installation/configuration/cmake_options.html#third-party-libraries-options), please ensure that your local dependencies are up-to-date.
```

## Namespace migrations and changes

The following list contains the namespace changes and migrations:

* All `eprosima::fastrtps::` namespace references were moved to `eprosima::fastdds::`.
* `SubscriptionBuiltinTopicData`, `PublicationBuiltinTopicData` and `ParticipantBuiltinTopicData` were moved from `fastdds::dds::builtin::` to `fastdds::dds::`.
* `EventKindBits::` references changed to`EventKind::`.
* `EventKindEntityId::` references changed to`EntityId::`.
* `StatisticsEventKind::` references changed to `statistics::EventKind::`.

> * `Duration_t` and `c_TimeInfinite` reference were moved to `dds::`.

Reviewer 2: ```suggestion
* `Duration_t` and `c_TimeInfinite` references were moved to `dds::`.
```

> * `ParticipantAttributes`, `PublisherAttributes`, `SubscriberAttributes`, `ReplierAttributes`, `RequesterAttributes` were moved to `xmlparser::`.

Reviewer 2: They are in fact private now. This list is missing `TopicAttributes`

Author: Ok so I will remove them from this list and I will keep only the public ones

> * `Time_t.hpp` references were explicitly distingueshed between `dds::Time_t::` and `rtps::Time_t::`.

Reviewer 2: We already had two classes, I think what we need to say is that they are in different files now

## Public headers migrated to *fastdds*

> All public header extensions have been changed to `.hpp`.

Reviewer 2: As this is the first time we say "public header", please clarify that those are the ones under `include`

The following list contains headers that were previously in the `include` folder and have been relocated to the `src/cpp` folder.
Since they are no longer public, it is not possible to include them in external projects:

* ParticipantAttributes.hpp
* ReplierAttributes.hpp
* RequesterAttributes.hpp
* PublisherAttributes.hpp
* SubscriberAttributes.hpp
* ProxyPool.hpp
* Semaphore.hpp
* MessageReceiver.hpp
* BuiltinProtocols.hpp
* shared_mutex.hpp
* StringMatching.hpp
* TimeConversion.hpp
* DBQueue.hpp
* ResourceEvent.hpp
* TimedEvent.hpp
* WriterProxyData.hpp
* ReaderProxyData.hpp
* ParticipantProxyData.hpp
* XML Parser API
* UnitsParser.hpp
* RTPSMessageGroup.hpp
* RTPSMessageCreator.hpp
* CDRMessage.hpp
* StatefulPersistentReader.hpp
* StatefulReader.hpp
* StatelessPersistentReader.hpp
* StatelessReader.hpp
* PersistentWriter.hpp
* StatefulPersistentWriter.hpp
* StatefulWriter.hpp
* StatelessPersistentWriter.hpp
* StatelessWriter.hpp
* logging.h
* Exception.h
* Cryptography.h
* Authentication.h
* AccessControl.h
* SecurityException.h
* ChangeForReader.hpp
* ReaderLocator.hpp
* ReaderProxy.hpp
* ServerAttributes.hpp
* TopicAttributes.hpp
* TypeLookupService.hpp

## API changes

> The table contains the list of API changes, showing the previous methods and the corresponding new ones introduced in Fast DDS v3.

Reviewer 2: ```suggestion
The table below contains the list of API changes, showing the previous methods and the corresponding new ones introduced in Fast DDS v3.
```
